Feature/ci template lint

fabmiz opened the pull request:
- Ticket: []
- Feature flag: n/a

## Purpose

Enable template linting on travis.
Currently, template lint is only run in a pre-commit hook that can be bypassed with
`--no-verify`

## Summary of Changes
Addressed:
- [`required-button-type`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/require-button-type.md)
- [`no-negated-condition`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-negated-condition.md)
- [`quotes`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/quotes.md)
- [`no-quoteless-attributes`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-quoteless-attributes.md)
- [`no-implicit-this`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-implicit-this.md)
- [`no-bare-strings`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-bare-strings.md)
- [`no-inline-styles`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-inline-styles.md)
- [`link-rel-noopener`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/link-rel-noopener.md)
- [`no-duplicate-attributes`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-duplicate-attributes.md)
- [`table-groups`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/table-groups.md)
-  [`link-href-attributes`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/link-href-attributes.md)
- Remove unnecessary `template-lint-ignore` *

## Side Effects

<!-- Any possible side effects? (https://en.wikipedia.org/wiki/Side_effect_%28computer_science%29) -->

## QA Notes
N/A Dev tested.

adlius commented on lib/analytics-page/addon/application/styles.scss:
```diff
@@ -53,3 +53,9 @@
+    padding: 0;
+    margin: 0;
+    border: 0;
+}
```
Why is this needed?

fabmiz replied:
Correct.

adlius commented on lib/app-components/addon/components/project-contributors/list/item/template.hbs:
```diff
@@ -107,9 +108,10 @@
+            type='button'
             class='btn btn-danger btn-sm'
             disabled={{not this.canRemove}}
+            {{action @removeContributor @contributor}}
```
Can we use element modifiers here?

aaxelb replied:
at risk of creeping scope, maybe should consider enabling [`no-action`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-action.md)?

fabmiz replied:
I think it's a good idea because of how confusing it's gotten. But as you say, let's defer enabling it to next iteration.

adlius commented on lib/app-components/addon/components/search-paginator/template.hbs:
```diff
@@ -3,9 +3,10 @@
                 local-class={{if (eq this.current item.text) 'active'}}
-                {{action (or item.action 'setPage') item.text}}
                 disabled={{item.disabled}}
+                {{action (or item.action 'setPage') item.text}}
```
Is it that element modifier doesn't take arguments?

fabmiz replied:
the `fn` helper can be used.

adlius commented on lib/app-components/addon/components/submit-section-buttons/template.hbs:
```diff
@@ -1,13 +1,24 @@
+        data-test-submit-section-discard
+        type='button'
+        class='btn btn-default'
+        {{action this.discard}}
```
Element modifier here?

adlius commented on lib/app-components/addon/components/submit-section-buttons/template.hbs:
```diff
@@ -1,13 +1,24 @@
+    type='button'
+    class='btn btn-primary'
+    disabled={{this.continueDisabled}}
+    {{action this.continue}}
```
Same here.

adlius commented on lib/app-components/addon/components/submit-section/complete/template.hbs:
```diff
@@ -1,18 +1,21 @@
+        <div
+            data-test-submit-section-click-to-edit
+            role='button'
+            {{action 'edit'}}
```
element modifier?

adlius commented on lib/collections/addon/components/collections-submission/template.hbs:
```diff
@@ -190,9 +192,10 @@
                         class='btn btn-success'
-                        {{action this.setShowSubmitModal}}
                         disabled={{not (eq this.activeSection this.sections.submit)}}
+                        {{action this.setShowSubmitModal}}
```
Same here.

adlius commented on translations/en-us.yml:
```diff
@@ -942,6 +942,7 @@ registries:
     discover:
         page_title: Search
         SHARE: SHARE
+        registrations: Registrations
```
Should this be in quotes?

fabmiz replied:
Quotes are generally not needed unless this is a multi-word string.

adlius replied:
Maybe we enforce the rule of always using quotes?

fabmiz replied:
This is beyond scope. Prefer we decide together as a team on this. Will make a note of this.

adlius commented on translations/en-us.yml:
```diff
@@ -1214,7 +1215,7 @@ registries:
     finalizeRegistrationModal:
         title: 'Almost done...'
-        notice: "Please keep in mind that:\n                <ul>\n                    <li>\n                        Registrations cannot be modified or deleted once completed.\n                    </li>\n                    <li>\n                        The content and version history of Wiki and OSF Storage will be copied to the registration.\n                    </li>\n                    <li>\ ...
+        notice: "Please keep in mind that:\n                <ul>\n                    <li>\n                        Registrations cannot be modified or deleted once completed.\n                    </li>\n                    <li>\n                        The content and version history of Wiki and OSF Storage will be copied to the registration.\n                    </li>\n                    <li>\ ...
```
Single quote?

fabmiz replied:
Because we have escapes `\`, YAML syntax requires us to use double quotes else a string `'\n'` will be returned instead of a newline.

adlius replied:
Ah... This is annoying. We could consider always using double quotes.

fabmiz replied:
This is beyond scope. Prefer we decide together as a team on this.

adlius commented on lib/collections/addon/components/discover-page/active-filters/filter/template.hbs:
```diff
@@ -8,6 +8,7 @@
+                type='button'
                 class='fa fa-times-circle'
                 local-class='removeActiveFilter'
                 {{action this.removeFilterItem}}
```
Could use element modifier here as well. But since you're not touching this line, probably not very important now.

adlius commented on lib/osf-components/addon/components/bs-alert/template.hbs:
```diff
@@ -1,6 +1,11 @@
+            type='button'
+            class='close'
+            aria-label={{t 'general.close'}}
+            {{action 'dismiss'}}
```
Same here.

adlius commented on lib/osf-components/addon/components/files/selected-list/template.hbs:
```diff
@@ -7,7 +7,7 @@
                 @items={{@selectedFiles}}
                 @use={{this.transition}}
-                @duration=500
+                @duration='500'
```
Got a question: for these type of parameters that expect integers/numbers, can we use `{{500}}`? `'500'` makes it look like a string.

adlius commented on lib/osf-components/addon/components/new-project-modal/template.hbs:
```diff
@@ -42,31 +42,33 @@
+                                            data-analytics-name='select_institution'
+                                            local-class='NewProject__institution-button'
+                                            @type='link'
+                                            @onClick={{action this.selectInstitution institution}}
```
Could use element modifier here.

fabmiz replied:
I think we want to use `on`  element modifier on HTML elements as opposed to components, because components may have their own way of handling the passed in `click` action.

aaxelb replied:
if the only thing the component is doing with the passed-in `onClick` is its own `{{on 'click' @onClick}}` on the element with `...attributes`, i think it'd be better to use `{{on}}` on the component -- it's nice to treat components the same as normal elements, when we can, and we could get rid of the `onClick` arg to simplify the component.

if there's a more complex component with a passed-in action that might be called for different/multiple reasons, then yeah, that should stay an argument -- but i'd probably want a name that indicates that it's more complex than "on click".

fabmiz replied:
I suggest we make this part of the ticket enabling the `no-action` rule. 

Element modifiers are not allowed in components with our version of ember-cli-template-lint. We need to upgrade `ember-cli-template-lint` to `^2.0.2`. `ember-cli-template-lint` is now deprecated and gets dropped in Ember 3.17 in favor `ember-template-lint`.

adlius commented on lib/osf-components/addon/components/osf-navbar/x-links/hyper-link/template.hbs:
```diff
@@ -7,8 +7,8 @@
             data-analytics-name='{{@analyticsLabel}}'
-            data-test-hyper-link=@testSelectorVal
-            class=@class
+            data-test-hyper-link={{@testSelectorVal}}
```
Does that mean this never worked before?

fabmiz replied:
🤷🏾‍♂️I don't think it worked, but also there seem to be no test looking for that test selector.

adlius commented on lib/osf-components/addon/components/sign-up-form/template.hbs:
```diff
@@ -52,33 +51,29 @@
-            }}
-            {{! template-lint-enable no-implicit-this }}
+                @model={{this.userRegistration}}
+                @valuePath='recaptchaResponse'
```
Can we always use `{{}}`? For example, can this be `{{'recaptchaResponse'}}`?

fabmiz replied:
`{{}}` are not needed here because `valuePath` is coerced to string. Basically, `{{}}` are needed for boolean and numbers so they don't get coerced to string unintentionally.

adlius replied:
I don't like how we have to think what kind of argument we are passing before choosing to use `{{}}` or not. If using `{{}}` on strings doesn't affect anything, might as well make it so that we always use `{{}}` regardless of argument types. We've seen enough confusion on when and when not to use `{{}}` already.

fabmiz replied:
This is beyond scope. Prefer we decide together as a team on this. Will make a note of this.

adlius commented on lib/registries/addon/components/registries-discover-search/x-sidebar/template.hbs:
```diff
@@ -1,18 +1,18 @@
-        <div class="col-xs-6">
-            <button data-test-clear-filters class="pull-right btn btn-sm btn-default" {{action 'clearFilters'}}>
+        <div class='col-xs-6'>
+            <button data-test-clear-filters type='button' class='pull-right btn btn-sm btn-default' {{action 'clearFilters'}}>
```
Split into multiple lines here?

adlius commented on lib/registries/addon/components/registries-discover-search/x-sidebar/template.hbs:
```diff
@@ -1,18 +1,18 @@
-    {{#each filters as |filter|}}
-        <button class="{{filter.class}}" local-class="Sidebar__ActiveFilter" {{action 'removeFilter' filter.filter}}>
+    {{#each this.filters as |filter|}}
+        <button type='button' class='{{filter.class}}' local-class='Sidebar__ActiveFilter' {{action 'removeFilter' filter.filter}}>
```
Same here.

adlius commented on lib/registries/addon/components/registries-header/template.hbs:
```diff
@@ -14,23 +13,36 @@
+                        <button
+                            class='btn btn-default'
+                            type='button'
+                            {{action 'toggleHelp'}}
```
Could use element modifier here.

adlius commented on lib/registries/addon/components/registries-registration-type-facet/template.hbs:
```diff
@@ -1,20 +1,19 @@
-{{#registries-search-facet-container title=title}}
-    <div local-class={{if onlyOSF 'RegistrationType' 'RegistrationType--Disabled'}}>
-        {{#unless onlyOSF}}
+{{#registries-search-facet-container title=this.title}}
```
Angle bracket invocations here?

adlius commented on lib/registries/addon/components/registries-search-result/template.hbs:
```diff
@@ -1,30 +1,30 @@
-                description=result.description
-                hyperlink=result.mainLink
-                resultId=result.id
+                title=@result.title
```
Angle bracket invocations here?

adlius commented on lib/registries/addon/components/registries-search-result/template.hbs:
```diff
@@ -36,61 +36,68 @@
+                aria-label={{t 'eosf.components.searchResult.showResult'}}
+                {{action this.toggleExpanded}}
+            >
+                {{fa-icon this.footerIcon size=1 ariaHidden=false}}
```
Angle bracket invocations here?

fabmiz replied:
This feels like scope creep. It's really related to the intent of this PR and without removing all the old invocation style, it does not help much. We could consider enabling [`no-curly-component-invocation`](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-curly-component-invocation.md) at some point?

aaxelb replied:
+1 for `no-curly-component-invocation` in another PR

adlius commented on lib/registries/addon/components/registries-services-list/template.hbs:
```diff
@@ -1,48 +1,50 @@
+                {{t 'registries.index.services.bottom.p1'}}
+                <div class='p-t-sm'>
+                    {{t 'registries.index.services.bottom.div.line1'}}
+                    <a local-class='ServicesList__ExternalLink' href='https://github.com/CenterForOpenScience/ember-osf-registries'
```
Multi-line?

adlius commented on lib/registries/addon/components/registries-services-list/template.hbs:
```diff
@@ -1,48 +1,50 @@
-                        onclick={{action "click" "link" "Index - public roadmap" "https://cos.io/our-products/product-roadmap/" target=analytics}}>
-                        {{~t "registries.index.services.bottom.div.linkText2"~}}
+                    {{t 'registries.index.services.bottom.div.line2'}}
+                    <a local-class='ServicesList__ExternalLink' href='https://cos.io/our-products/product-roadmap/'
```
Multi-line?

adlius commented on lib/registries/addon/components/registries-services-list/template.hbs:
```diff
@@ -1,48 +1,50 @@
                 </div>
             </p>
-            <a href="mailto:contact@osf.io" class="btn btn-info btn-lg" onclick={{action "click" "link" "Index - Contact" target=analytics}}>{{t "registries.index.services.bottom.contact"}}</a>
+            <a href='mailto:contact@osf.io' class='btn btn-info btn-lg' onclick={{action 'click' 'link' 'Index - Contact' target=this.analytics}}>
```
Same here.

adlius commented on lib/registries/addon/discover/template.hbs:
```diff
@@ -1,16 +1,21 @@
 {{title (t 'registries.discover.page_title')}}
 
-{{#registries-header showHelp=true value=(mut query) onSearch=(action 'onSearch') searchable=searchable as |header|}}
+{{#registries-header showHelp=true value=(mut this.query) onSearch=(action 'onSearch') searchable=this.searchable as |header|}}
```
Angle bracket invocation here?

fabmiz commented:
Addressed most comments, but in the spirit of keeping the changes close to the intention of the PR (turn on template lint on CI i.e fixing existing rule violations), additional work (enabling new rules) is deferred.
